File: C++/代码随想录/Binary_Tree/Modification and construction of Binary tree/226. Invert Binary Tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root


# Test case
root = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7, TreeNode(6), TreeNode(9)))
sol = Solution()
inverted_tree = sol.invertTree(root)

# 打印用例
def inorder_traversal(root):
    return inorder_traversal(root.left) + [root.val] + inorder_traversal(root.right) if root else []
print(inorder_traversal(inverted_tree))  # [9, 7, 6, 4, 3, 2, 1]

File: C++/代码随想录/Binary_Tree/Modification and construction of Binary tree/226. Invert Binary Tree.cpp
// 226. Invert Binary Tree
// https://leetcode.com/problems/invert-binary-tree/
// clang++ 226.\ Invert\ Binary\ Tree.cpp -std=c++17
#include "../../printUtils.hpp"

template <typename T>
TreeNode<T>* invertTree(TreeNode<T>* root) {
    if (root == nullptr) return nullptr;

    TreeNode<T>* left = invertTree(root->left);
    TreeNode<T>* right = invertTree(root->right);

    root->left = right;
    root->right = left;
    
    return root;
}

int main() {
    TreeNode<int>* root = buildTree({4, 2, 7, 1, 3, 6, 9}, -1);
    TreeNode<int>* inverted = invertTree(root);

    std::cout << "Inverted binary tree:" << std::endl;
    printArray(levelOrder(inverted));

    return 0;
}